### PR TITLE
Pagination refactor and feed sort by latest feed version import

### LINKED ISSUE #823: Feed: sort by latest feed version import
Make feeds sortable by the date of the most recent import attempt

### PR DESCRIPTION
**Pull request by irees**
Resolves #823

### CODE REVIEW

File: app/controllers/concerns/json_collection_pagination.rb [hunk 1]
```diff
@@ -4,56 +4,47 @@ module JsonCollectionPagination
...
+    params[:sort_key] = sort_key
+    params[:sort_order] = sort_order
+    params[:offset] = offset
+    params[:per_page] = per_page
+    params[:total] = total
+    self.params = params
+    paginated_json_collection_new(collection)
+  end
 
+  def paginated_json_collection_new(collection)
```
Comment (irees) [inline app/controllers/concerns/json_collection_pagination.rb]: rename

Comment (drewda) [inline app/controllers/concerns/json_collection_pagination.rb]: Or simpler in the long run to just go through and update all the existing calls to `paginated_json_collection`?

File: app/controllers/concerns/json_collection_pagination.rb [hunk 2]
```diff
@@ -4,56 +4,47 @@ module JsonCollectionPagination
...
-      # worst case: the query takes longer than 2 minutes and is
-      # killed at the database and the load balancer
-      meta[:per_page] = per_page
-      data_on_page = collection
-    else
-      per_page = (per_page.presence || self.class::PER_PAGE).to_i
-      meta[:per_page] = per_page
-      # Get the current page of results.
-      #  Add +1 to limit to see if there is a next page.
-      #  This will be dropped in the return.
```
Comment (drewda) [inline app/controllers/concerns/json_collection_pagination.rb]: I think this comment is worth retaining to explain why there's that `per_page+1`

File: app/controllers/concerns/json_collection_pagination.rb [hunk 3]
```diff
@@ -67,4 +58,38 @@ def paginated_json_collection(collection, path_helper, sort_key, sort_order, off
...
+    fail ArgumentError.new('Invalid sort_key') unless collection.column_names.include?(key.to_s)
+    collection.reorder(key => sort_order)
+  end
+
+  def sort_offset
+    (params[:offset].presence || 0).to_i
+  end
+
+  def sort_per_page
+    return false if ['false', false, '∞'].include?(params[:per_page])
```
Comment (drewda) [inline app/controllers/concerns/json_collection_pagination.rb]: Took me a while to trace through how this line of logic is still able to disable `limit`ing the results query. Would it be easy to understand if it were up in `paginated_json_collection_new` or in a differently named method?

File: spec/controllers/json_collection_pagination_spec.rb [hunk 4]
```diff
@@ -119,7 +122,7 @@ class FakeController < ApplicationController
     it 'allows pagination to be disabled' do
       data = object.send(:paginated_json_collection, collection, path_helper, 'id', 'desc', 0, '∞', false, {})
       expect(data[:json].items.count).to eq 15
-      expect(data[:meta][:per_page]).to eq '∞'
+      expect(data[:meta][:per_page]).to eq false
```
Comment (drewda) [inline spec/controllers/json_collection_pagination_spec.rb]: I think Dispatcher and some other clients are expecting the response of  '∞'. Would it be possible to keep that unchanged?

Comment (irees) [inline spec/controllers/json_collection_pagination_spec.rb]: Yes, will change.

File: app/controllers/concerns/json_collection_pagination.rb (hunk 2 above)
Comment (irees) [inline app/controllers/concerns/json_collection_pagination.rb]: ok

**Pull request merged**